pre_commit/languages/download.py

```diff
@@ -0,0 +1,291 @@
+        return os
+
+    @property
+    def cpu(self) -> str:
+        _, cpu = self.parts
+        return cpu
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @staticmethod
+    def host() -> Platform:
+        os = Platform.OS[platform.system()]
+        cpu = Platform.CPU[platform.machine()]
+        return Platform(f'{os}/{cpu}')
+
+    def __str__(self) -> str:
+        return f'{self.os}/{self.cpu}'
+
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, Platform):
+            return False
+        return other.os == self.os and other.cpu == self.cpu
+
+
+class ChecksumMismatchError(ValueError):
+    def __init__(
+        self, *, expected: str, actual: str,
+        message: str = 'checksum mismatch',
+    ):
+        self.__expected = expected
+        self.__actual = actual
+        self.__message = message
+
+    @property
+    def expected(self) -> str:
+        return self.__expected
+
+    @property
+    def actual(self) -> str:
+        return self.__actual
+
+    @property
+    def message(self) -> str:
+        return self.__message
+
+    def __str__(self) -> str:
+        return f'''{self.__message}:
+ - expected: {self.expected}
+ - actual  : {self.actual}
+'''
+
+
+class HasBinaryRead(Protocol):
+    def read(self, __size: int = -1) -> bytes | None: ...
+
+
+class SRI:
+    def __init__(self, value: str):
+        self.__value = value
+        self.__algorithm, self.__checksum = self.value.split('-', 1)
+        if self.algorithm not in hashlib.algorithms_available:
+            raise ValueError(f"`{self.algorithm}` is not available, \
+choose one of: {','.join(hashlib.algorithms_available)}`")
+
+        if b64encode(b64decode(self.checksum)).decode('utf-8') !=\
+                self.checksum:
+            raise ValueError('Invalid checksum string, \
+the checksum string has to be encoded in base64.')
+
+        hasher = hashlib.new(self.algorithm)
+
+        checksum_len = len(b64decode(self.checksum))
+        if checksum_len != hasher.digest_size:
+            raise ValueError(
+                f'Invalid checksum string length of {checksum_len} for \
+{self.algorithm}, expected {hasher.digest_size}',
+            )
+
+    @property
+    def value(self) -> str:
+        return self.__value
+
+    @property
+    def algorithm(self) -> str:
+        return self.__algorithm
+
+    @property
+    def checksum(self) -> str:
+        return self.__checksum
+
+    def __str__(self) -> str:
+        return self.value
+
+    def check(self, io: HasBinaryRead, chunk: int = 4096) -> Iterator[bytes]:
+        hasher = hashlib.new(self.algorithm)
+        while buffer := io.read(chunk):
+            hasher.update(buffer)
+            yield buffer
+        digest = b64encode(hasher.digest()).decode('utf-8)')
+        if digest != self.checksum:
+            raise ChecksumMismatchError(
+                expected=self.checksum,
+                actual=digest,
```

Comment:
It might be nice to provide the full SRI so that the user can pass that into local tools to verify (such as `srisum`).

```suggestion
                expected=self,
                actual=f"{self.algorithm}-{digest}",
```

Reply:
Done